### ISSUE TITLE
RebootCommand option in sddm.conf seems ignored

### ISSUE THREAD

**Issue by egreshko** [sddm/sddm#434]
There is a problem when shutting down a system which uses Wifi as its network connection and has NFS mounts.  The network is dropped before unmounts so if you try to do a normal "systemctl poweroff" the system will take a very long time to shutdown.  So, one must unmount the NFS mounts first.  Running this script from the command line works well.

#!/bin/sh
/usr/bin/umount -a -t nfs4 
/usr/bin/systemctl reboot

So, I placed the script in /usr/sbin and modified sddm.conf to contain

RebootCommand=/usr/sbin/rebooty

However, clicking on "Reboot" from the login screen seems to have no effect.  The system does not unmount the NFS mounts and reboot takes forever.  

For completeness I'm running Fedora F22 with sddm-0.10.0-4 I have filed a BZ with Fedora but the suggestions haven't helped.  https://bugzilla.redhat.com/show_bug.cgi?id=1231413

**Comment by plfiorini** [comment]
Proposal:
* RebootCommand is empty by default
* login1 backend will use RebootCommand if specified otherwise the D-Bus request
* upower backend will use RebootCommand (assuming "reboot" if it's empty)

**Comment by jleclanche** [comment]
Do we do this for 0.12? If we dont, we have to document that RebootCommand isn't used in login1.

**Comment by jleclanche** [comment]
I know we just said on IRC we can release without it, but I'm starting to think we should do this for 0.12.

**Comment by giddie** [comment]
Any progress on this? I'm still seeing this issue in 0.13 (ArchLinux)

**Comment by flyser** [comment]
still present.

**Comment by giddie** [comment]
Still broken :(

**Comment by felagund** [comment]
Indeed. I wanted to set the reboot option to `/bin/systemctl kexec` but it just does not seem to register.

**Comment by iam-TJ** [comment]
The option `RebootCommand` in `sddm.conf` is only valid when the `UPower` backend is being used, and that backend is only used if both DBus services `login1` and `ConsoleKit` are not found at run-time.

The best solution, instead of mixing up `RebootCommand` may be to add `BackendPreferred=` to `sddm.conf` with options `login1|ConsoleKit|UPower`.

In which case instead of the first available backend being used to trigger the reboot `BackendPreferred` would be chosen if available.

Then for users that expect `RebootCommand=` to be used they can set `BackendPreferred=UPower`. There'd need to be a log entry confirming (or not) that the preferred back-end is available at run-time.

With clear documentation in the example `sddm.conf`, man-page and help, it should resolve much frustration and confusion.

On installations that use `systemd`, and specifically `systemd-logind`, the DBus `org.freedesktop.login1.Manager` interface is used by `sddm` to trigger the restart by (currently) calling `Reboot()`. I've just finished a patch that uses `RebootWithFlags(SD_LOGIND_KEXEC_REBOOT)` instead. Provided a `kexec` kernel, initialramfs, and command-line have been loaded, the system reboots using it.
```
$ sudo kexec -l /vmlinuz --initrd /initrd.img --command-line="root=/dev/mapper/SUNNY-rootfs ro systemd.unified_cgroup_hierarchy=1 radeon.drm=1 nosplash
$ cat /sys/kernel/kexec_loaded
1
```
```systemd-logind[1942]: System is rebooting with kexec.
$ journalctl -b -2 -u sddm
systemd[1]: Started Simple Desktop Display Manager.
sddm[8549]: Initializing...
sddm[8549]: sddm QDBusInterface service = "org.freedesktop.login1" path = "/org/freedesktop/login1" interface = "org.freedesktop.login1.Manager"
sddm[8549]: sddm QDBusInterface service = "org.freedesktop.UPower" path = "/org/freedesktop/UPower" interface = "org.freedesktop.UPower"
sddm[8549]: Starting...
sddm[8549]: Logind interface found
...
sddm[8549]: Greeter session started successfully
sddm[8549]: Message received from greeter: Connect
sddm[8549]: Message received from greeter: Reboot
sddm[8549]: sddm reboot using "org.freedesktop.login1"
sddm[8549]: sddm calling RebootWithFlags
systemd-logind[1942]: System is rebooting with kexec.
...
```

**Comment by felagund** [comment]
Why is the proposal [above](https://github.com/sddm/sddm/issues/434#issuecomment-121978530) not possible? Not all of these backends can use kexec?  

> * RebootCommand is empty by default
> * login1 backend will use RebootCommand if specified otherwise the D-Bus request
> * upower backend will use RebootCommand (assuming "reboot" if it's empty)

If I understand it correctly, `SDDM` will from now on reboot using kexec when the previous use was done with kexec - but as of now, with `login1` and `UPower`, it cannot initiate it on its own, right?

**Comment by Vogtinator** [comment]
IMO `RebootCommand` does not make sense on systems which have login1. If login1 can't be used to reboot, that needs to be fixed.

If NFS mounts are somehow unmounted at the wrong moment, dependencies are set up incorrectly. Either `network-online.target` is broken or the mounts don't depend on it.

**Comment by tvidal-net** [comment]
I am also being affected by this.

> IMO `RebootCommand` does not make sense on systems which have login1. If login1 can't be used to reboot, that needs to be fixed.

While I partially with you, at the very least the `sddm.conf` man page should be updated to inform in which conditions `RebootCommand` is honoured or ignored in case the *default behaviour* is not desired. If it does not behave like it is documented, it should be documented how it actually behaves.

IMO, the better option would be to execute whatever is defined in `RebootCommand` IF `RebootCommand` is present in the config, regardless of the backend being used, sort of an "override" or maybe have an alternative `RebootCommandOverride` key to avoid breaking current behaviour of ignoring `RebootCommand`. An extra feature would be to disable reboot entirely if `RebootCommand=` is present with an empty value in the config.

What I am actually trying to do, is to boot on a secondary OS on a dual-boot installation on a touch-screen device with no physical keyboard (Surface Pro) as soon as possible. Since there is no keyboard to access the firmware setup and change the boot option, having sddm to immediately run `efibootmgr --bootnext` before reboot would solve my issue, had `RebootCommand` be implemented like it is documented on the manual page.

**Comment by grahamperrin** [comment]
Cross-reference: 

* FreeBSD bug [270835 – x11/sddm: No shutdown and reboot buttons without sysutils/consolekit2](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=270835)

– [**comment 8**](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=270835#c8) in particular. 

----

(Ignore my comments 1, 2, 5 and 7. Ahem.)
